Implement SmptAuthEmailSender and GmailSmtpEmailSender

selo-ecube opened the pull request:
There have been two modifications to the EmailSender interface:
(1) I introduced a new functional interface MessageEditor. It sounds like the
    MessageAuthor that we removed before and it is indeed quite similar with 1
    modification: it declares Exception so it can pass on all checked exceptions
    that get thrown during message creation - particularly MessagingException
    which is thrown by all(?) methods of MimeMessage. Please feel free to
    propose an alternative solution or an alternative name.
(2) Also due to the issue with MimeMessage-methods throwing MessagingException,
    EmailSender.send(...) now throws EmailSenderException (e.g. those that
    wrap MessagingException during message creation). This seems necessary
    because the message shall be created before EmailSender.send(...) returns,
    i.e. these exceptions cannot(?) be wrapped inside the CompletionStage that
    is used for sending the email.

I decided to make all internal methods of SmtpAuthEmailSender protected because
I myself never like to come across classes whose authors could not think of any
reason to subclass them. One particular reason to subclass SmtpAuthEmailSender
and to override SmtpAuthEmailSender.createProperties(...) is that there are many
configuration properties in the Java Mail API that make it likely that the
current implementation does not fit all SMTP servers in the world.

In the JavaDoc of SmtpAuthEmailSender.createProperties(...) I wrote "Consult the
source code of this method before you override it" instead of listing the
properties the method sets. I'm not sure whether you deploy the source code in
Maven along with the binaries. I thought not replicating the properties in the
JavaDoc might make it less likely that the JavaDoc becomes outdated.

Please review, @lauraluiz and @schleichardt !

selo-ecube committed:
Implement SmptAuthEmailSender and GmailSmtpEmailSender

There have been two modifications to the EmailSender interface:
(1) I introduced a new functional interface MessageEditor. It sounds like the
    MessageAuthor that we removed before and it is indeed quite similar with 1
    modification: it declares Exception so it can pass on all checked exceptions
    that get thrown during message creation - particularly MessagingException
    which is thrown by all(?) methods of MimeMessage. Please feel free to
    propose an alternative solution or an alternative name.
(2) Also due to the issue with MimeMessage-methods throwing MessagingException,
    EmailSender.send(...) now throws EmailSenderException (e.g. those that
    wrap MessagingException during message creation). This seems necessary
    because the message shall be created before EmailSender.send(...) returns,
    i.e. these exceptions cannot(?) be wrapped inside the CompletionStage that
    is used for sending the email.

I decided to make all internal methods of SmtpAuthEmailSender protected because
I myself never like to come across classes whose authors could not think of any
reason to subclass them. One particular reason to subclass SmtpAuthEmailSender
and to override SmtpAuthEmailSender.createProperties(...) is that there are many
configuration properties in the Java Mail API that make it likely that the
current implementation does not fit all SMTP servers in the world.

In the JavaDoc of SmtpAuthEmailSender.createProperties(...) I wrote "Consult the
source code of this method before you override it" instead of listing the
properties the method sets. I'm not sure whether you deploy the source code in
Maven along with the binaries. I thought not replicating the properties in the
JavaDoc might make it less likely that the JavaDoc becomes outdated.

schleichardt commented on email-smtp/it/AsynchronousIntegrationTest.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    @Test
+    public void sendTwoMessagesAndWaitForBoth() {
```
I think it is not necessary to guarantee the order of the messages. Some servers are slower than others and emails can differ in sizes so it is reasonable that a later send email still can arrive before the first.

schleichardt commented on email-smtp/src/main/java/io/commercetools/sunrise/email/smtp/SmtpAuthEmailSender.java:
```diff
@@ -1,27 +1,242 @@
+     * @param timeoutMs         the timeout for creating, reading from and writing to SMTP connections in
+     *                          milliseconds, see above for details
+     */
+    public SmtpAuthEmailSender(@Nonnull final Executor executor, @Nonnull final String host, final int port,
```
It would be very useful to be able to provide a properties parameter here to be able to override stuff which is set in createProperties.

```
final Properties properties = createProperties(host, port, transportSecurity, timeoutMs);
// here override all properties in properties with the constructor parameter.
```

schleichardt replied:
There is the possibility to override the `createProperties` method, but it contains a lot of parameters and possible in the future more. So it could make sense to make `createProperties` private and add sth like

``` java
protected void properties(final Properties properties) {
}
```

which can be overridden like:

``` java
protected void properties(final Properties properties) {
    properties.setProperty("mail.smtp.ssl.checkserveridentity", "" + false);
}
```

selo-ecube replied:
Which of the two options do you prefer?
I'm in favour of the second one (adding properties(Properties) ) because it separates passing values for the default properties configuration from adding additional properties.

lauraluiz replied:
Focusing on the constructor (so I'm not touching the same topic as @schleichardt , which was aimed for subclasses, instead I'm just a user of the `SmtpAuthEmailSender`) we are mixing two type of parameters:
- SMTP related parameters (i.e. `host`, `port`, `transportSecurity`, `username`, `password`), which might also grow in the future, and
- Fine-tuning parameters, such as `executor` and `timeoutMs`, which for the getting started might be a bit overwhelming.

I would try to organize them differently. So for the first case I would use a Class to encapsulate the proper SMTP configuration, which might follow the Builder pattern to enable expansion in the future. The reason why I would not use the `Properties` instance is because I don't want developers to know the property key they have to use e.g. for host `mail.smtp.host`. Of course internally we will use the Properties class and with the `createProperties` and `properties` methods, but not from outside.

For the second topic, if we use a config class for the rest, then these two factors (timeout and executor) can go at the end together. I would think it's positive if we provide a default constructor with these two values already defined for the fast getting started situation, some values that can generally give an acceptable result before the fine-tuning.
I hope I explained myself properly, what do you think @schleichardt @selo-ecube ?

selo-ecube replied:
I agree that the constructor parameters may be grouped into two concerns. Re-arranging constructor arguments makes good sense, i.e. to have `executor` come after `password` and before `timeoutMs`.

I'd rather not want to modify the constructor to fit anticipated future use cases before they have arrived. I.e. I'd refrain from introducing a builder for the SMTP-related constructor arguments. That would create a second way of modify the configuration besides overriding SMTPAuthEmailSender.properties(...) - before we know which properties/parameters actual users of the API need to add.

I could add a constructor with default values for executor and timeoutMs. I'm personally rather not a fan of providing a fast start because fast solutions typically overlook corner cases and require time investment later on when you though you were done already. But whether or not fast startup shall be supported is a decision that you need to make. If I were to choose defaults, I'd use ForkJoinPool.commonPool() as default executor and a timeout of 1000 ms - the latter one being a mere guess that does not consider any context in which the email sender might be used.

lauraluiz replied:
Another solution with less impact is to put the suggestions/example in the documentation, just as a guidance for the indecisive.

Regarding the constructor, it doesn't matter whether it's a Builder or not, as currently all the parameters are mandatory. It can perfectly be just a class that contains the parameters that we are currently passing. But having 7 parameters in the constructor is a sign that maybe we are mixing too many concerns.

selo-ecube replied:
I'll improve the JavaDoc of the timeoutMs parameter (the JavaDoc for executor already suggests a default) and add an example of how to invoke the constructor in the constructor JavaDoc.

I'll add a SmtpConfiguration class an instance of which is passed to the SmtpAuthEmailSender constructor. SmtpConfiguration will be immutable, with a constructor taking all SMTP-related parameters that were previously passed to the SmtpAuthEmailSender constuctor.

schleichardt commented on email-smtp/it/ConfigurationIntegrationTest.java:
```diff
@@ -0,0 +1,107 @@
+            try {
+                io();
+            } catch (IOException e) {
+                throw new IllegalStateException("I/O failed", e);
```
Do not change this here.

Normally https://docs.oracle.com/javase/8/docs/api/java/io/UncheckedIOException.html can be used to get an unchecked IOException.

schleichardt commented on email-smtp/src/main/java/io/commercetools/sunrise/email/smtp/SmtpAuthEmailSender.java:
```diff
@@ -1,27 +1,242 @@
-        return CompletableFuture.supplyAsync(() -> "");
+    public CompletionStage<String> send(@Nonnull final MessageEditor messageEditor) {
+        final MimeMessage message = createAndFillMessage(messageEditor);
+        return CompletableFuture.supplyAsync(() -> {
```
This way you can also put checked exceptions into the future:

``` java
    @Override
    @Nonnull
    public CompletionStage<String> send(@Nonnull final MessageEditor messageEditor) {
        final MimeMessage message = createAndFillMessage(messageEditor);
        final CompletableFuture<String> result = new CompletableFuture<>();
        executor.run(() ->{
           try {
                sendMessage(message);
                result.complete(message.getMessageID());
            } catch (final Throwable t) {//IDE may warn about this, but if fatals are not in the result, it might hang forever
                result.completeExceptionally(t);
            }
        });
    return result;
    }
```

selo-ecube replied:
I didn't know this option, thank you! 
Do you think it would be the way to go in our case? (Given that we specified that the CompletionStage should contain an EmailSenderException wrapping the original exception.)

schleichardt replied:
Yes.

The Java 8 async stuff isn't exactly well designed and documented.
I think it is way to go since it is not so restrictive concerning the error handling and transportation.

schleichardt commented:
Great.

The MessageEditor is a good solution for the problem.

For the CompletionStage and exception problem I made a suggestion in https://github.com/commercetools/commercetools-sunrise-java-email/pull/5#discussion_r64532197 .

There are some comments you could address.

It is okay and correct to link to the properties, we can't maintain a complete list.
It is a Maven Central requirement to also publish the Javadoc and the sources.

selo-ecube commented:
Thanks for the review! I'd wait until Laura's review so I can address all comments at once, if you don't mind. (I'll be out of office Thursday and Friday this week, back again next Monday.)

schleichardt commented:
okay

lauraluiz commented on email-common/src/main/java/io/commercetools/sunrise/email/MessageEditor.java:
```diff
@@ -0,0 +1,21 @@
+ * {@link MessagingException}s thrown by {@link MimeMessage}.
+ */
+@FunctionalInterface
+public interface MessageEditor {
```
This one makes sense, so we should keep it. But we still have the problem that if you don't read the documentation (which of course shouldn't be the case, but we know lazy developers ;)) you will lost without a clue of what to do with `MessageEditor`, which creates frustration and that's the worst first experience when you work with something new.
To minimize that, what if we provide a basic implementation of `MessageEditor` which expects the typical information for an email (i.e. from, recipients, subject, content). People looking for implementations of `MessageEditor` will find it and use it, once they realize they are missing something more advanced, they will think about creating their own implementation.
The pros are clear: easier getting started (easier to achieve basic usage), more intuitive, less fully relying on Java 8 knowledge (better for beginners)... cons are that we would be leaving as a background solution what is actually quite powerful itself and still simple (once you know it).
An alternative solution would be to document in this interface that you are supposed to implement it yourself and an example. People not knowing what to do will likely end up in this file looking for a clue. Actually I would do it nonetheless, also with the basic implementation solution.
@schleichardt @selo-ecube what do you think about it?

lauraluiz replied:
Regarding the example, we can also redirect to the `send` method Javadoc, to avoid repetition ;)

schleichardt replied:
So it would be easier to write a POJO which can be converted to a MimeMessage.

It sounds for me like we should have used Apache Email which already has that kind of bean.

lauraluiz replied:
But Apache Email had a `send` method I think, didn't it?

lauraluiz replied:
I mean, Apache Email is both the EmailSender and the Message. That can create great confusion on where to put what. That's why we chose MimeMessage, because it is purely information about the Message, which is what we need.

schleichardt replied:
It is worse, it has multiple send methods: http://commons.apache.org/proper/commons-email/javadocs/api-1.3.3/org/apache/commons/mail/Email.html

selo-ecube replied:
I would guess that creating a POJO that can be converted to MimeMessage takes a considerable amount of time given how many methods the MimeMessage class of the Java Mail API and the related classes in Apache Commons Email have.

Given the drawbacks of Apache Commons Email mixing multiple responsibilities in the Email class, would you be fine with improving the JavaDoc of MessageEditor, as Laura suggested?

schleichardt replied:
ok

schleichardt replied:
@lauraluiz do we here also need test code inclusion like in the JVM SDK or just write examples into the Javadoc?

lauraluiz replied:
I would go for plain examples, since the components of this module are quite small to forget to update the documentation, and shouldn't be changing much. So if we change the Interface itself, we only have to take care of two classes. For any implementation, it's only the class itself.

lauraluiz commented on email-common/src/main/java/io/commercetools/sunrise/email/EmailSender.java:
```diff
@@ -44,15 +43,19 @@
+     *                      Note that {@link MimeMessage} instances are not immutable. Messages passed to the editor
+     *                      must only be used by that editor instance and must not be passed elsewhere.
      * @return A completion stage for sending the message asynchronously.
+     * @throws EmailSenderException if there was an error while creating or filling the message. Note that in contrast
```
I would create 2 different exceptions to avoid confusion, one for the email creation and another for the email sending. Email creation exception will be directly thrown and the Email sending exception will be thrown wrapped in the `CompletionStage`. That way we can have two different `@throws` documentation and be clearer about it.

selo-ecube replied:
Ok, I'll add an EmailCreationException.

(I wouldn't add two @throws tags in the method JavaDoc because only EmailCreationException would be thrown and EmailSenderException might be contained in the CompletionStage returned by EmailSender.send(...), but not strictly be thrown to callers of that method.)

lauraluiz replied:
With the way @schleichardt showed you to wrap any exception inside the CompletionStage, let's go for it. The separation of the two exceptions is still a good idea, we can even create a superclass exception that contains both of them. So EmailSenderException will have two subclasses: EmailCreationException and EmailSendingException (I don't like that the name is so similar to Sender, any better idea?). This way developers can still catch any exception the same way.

lauraluiz replied:
Transmission? Dispatching? Transference?

selo-ecube replied:
The hierarchy `EmailCreationException extends EmailSenderException` and `EmailDeliveryException extends EmailSenderException` sounds good.

I'd choose `EmailDeliveryException` over `EmailTransmissionException` if there are objections.

> With the way @schleichardt showed you to wrap any exception inside the CompletionStage, let's go for it.

Just to make sure my understanding is correct: Do you mean that EmailSender.send(...) should neither throw EmailCreationException, nor EmailDeliveryException but instead wrap both kinds of exception in the returned CompletionStage?

lauraluiz replied:
Delivery, that definitely sounds nicer! 👍 

So the resulting exception, after a short discussion, ends up:
- EmailCreation, as a non-checked exception thrown normally, and documented normally in the `@throw`.
- EmailDelivery, as a non-checked exception wrapped in the CompletionStage, documented in the `@result`.
  I personally do not like the documentation part, because Javadoc should help developers to easily see anything they have to take care of and result is probably not the best place to see it, but that's all the tools that we have now.

As code says more than words, if anybody would like to catch everything, they should do:

``` Java
try {
    emailSender.send(...)
        .exceptionally(throwable -> {
            if (throwable.getCause() instanceof EmailDeliveryException) {
                // do something
            }
            ...
        });
} catch (EmailCreationException e) {
    // do something
}
```

selo-ecube commented:
I added my 2 cents where I thought it might be beneficial. I don't want to imply a need to rush, though.

lauraluiz commented:
OK, I've finally finished reviewing on my side.

selo-ecube committed:
Implement review comments for https://github.com/commercetools/commercetools-sunrise-java-email/pull/5

selo-ecube committed:
Fix JavaDoc error

selo-ecube commented:
Thanks for the input, @lauraluiz and @schleichardt ! Please review the modifications!

lauraluiz commented on email-common/src/main/java/io/commercetools/sunrise/email/EmailSender.java:
```diff
@@ -41,19 +41,43 @@
      * @param messageEditor the email sender passes an empty message to the message editor that the editor shall fill.
      *                      Note that {@link MimeMessage} instances are not immutable. Messages passed to the editor
      *                      must only be used by that editor instance and must not be passed elsewhere.
      * @return A completion stage for sending the message asynchronously.
```
I would put in the return:
`@return A completion stage containing a string identifying the delivered email, or a CompletionException with cause {@link EmailDeliveryException} if the email could not be sent successfully`

lauraluiz replied:
We can also modify a bit the sentence to make it clear that EmailDeliveryException is not the only one.

selo-ecube committed:
Adjust exception documentation for EmailSender.send(...)

The example in EmailSender.send(...) did not correctly represent that the EmailDeliveryException is
now contained in the CompletionStage directly without being wrapped in a CompletionException. This
change occurred when we switched from using CompletableFuture.supplyAsync(...) to Executor.exec(...)
and the direct use of CompletionStage.execeptionally(...).

I also took up a change in the @return JavaDoc that Laura proposed.
